Add expired event

aljones15 opened the pull request:
if `ttl` is set in the session on refresh, then we call `_setTimeout` that emits a new event `expire` when the `ttl` is reached. The app can then interpret this `expire` event as needed.


Please note I have compiled a gist with all the behaviors I observed from various settings: https://gist.github.com/aljones15/3defa74231551e5b9934d79609661ae1

It might help.

Addresses: https://github.com/digitalbazaar/bedrock-web-session/issues/10

Remaining issues:
- [ ] github actions is extremely flaky with this branch
- [x] add localStorage for the ttl so tabs can communicate with each other
- [ ] add an actual test for the expire event

aljones15 committed:
Add an expire event.

aljones15 committed:
Implement setTimeout.

  - add this._timeout to constructor
  - add expire event to event types

dlongley commented on Session.js:
```diff
@@ -44,7 +53,23 @@ export default class Session {
+   *
+   * @returns {undefined} Just emits an event.
+  */
+  setTimeout({timeout}) {
```
I don't think this should be a public method on the session.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-session/commit/3ba295a12e0a1309909774da1a5da905d965e1ef

dlongley commented on Session.js:
```diff
@@ -20,12 +25,16 @@ export default class Session {
   async refresh({authentication} = {}) {
     const oldData = this.data;
     const newData = await this._service.get();
+    if(typeof newData.originalMaxAge === 'number') {
```
I'm assuming this will change to use `newData.ttl` instead.

aljones15 replied:
we do need a way to set the timeout using expires in case someone does want to use a timestamp over ttl. I went ahead and changed to `ttl` and made `_setTimeout` private. https://github.com/digitalbazaar/bedrock-web-session/pull/5/commits/3ba295a12e0a1309909774da1a5da905d965e1ef

dlongley commented on main.js:
```diff
@@ -17,6 +17,7 @@ export const createSession = async (
 
 export const getSession = async (
   {id = 'session.default', store = defaultStore} = {}) => {
+  // note: you might need to refresh this.
```
```suggestion
  // the cached session is simply returned and intentionally not refreshed here;
  // if the caller wants the session refreshed they may call `refresh()` on the result
```

aljones15 committed:
Update Comment on cached session.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

aljones15 committed:
Lint long comment in main.

aljones15 committed:
Make setTimeout a private method & use ttl.

aljones15 committed:
Add session http & mongodb to test deps.

aljones15 committed:
Add test for unauthenticated request.

aljones15 committed:
Add tests for create and refresh session.

aljones15 committed:
Add deps.

aljones15 committed:
Add an unauthenticated test.

aljones15 committed:
Add authenticated session test.

aljones15 committed:
Add a logout test.

aljones15 committed:
Remove get API test suite.

aljones15 committed:
Add express session settings.

aljones15 committed:
Make test sessions expire after 1 second.

aljones15 committed:
Add tests for refresh.

aljones15 committed:
Add session tests.

aljones15 committed:
Add createSession test.

aljones15 committed:
Add more assertions on session.data.

aljones15 committed:
Add change event tests.

aljones15 committed:
Add copyrights to session tests and helpers.

aljones15 committed:
Remove redudant comments.

aljones15 committed:
Use eql shorthand for deep.equal.

Co-authored-by: mattcollier <902041+mattcollier@users.noreply.github.com>

aljones15 committed:
Scope session to each test.

aljones15 committed:
Add comments to explain refresh tests.

  - Use eql over deep.equal

aljones15 committed:
createAccount only returns account & totp.

aljones15 committed:
Add try catch to promises to ensure no unhandled rejections occur in tests.

aljones15 committed:
Ensure logout before account creation.

aljones15 committed:
Use should.have.keys over Object.keys

  - Reuse the same session id for session.js tests
  - Clean up session store after tests

aljones15 committed:
Throw if duplicate account in createAccount.

aljones15 committed:
Add a delay to reduce CI flakiness.

aljones15 committed:
Run tests in new windows not iFrames.

aljones15 committed:
Try with concurrency set to 1.

aljones15 committed:
Try using iframe again.

aljones15 committed:
Remove new window code.

aljones15 committed:
Remove profile tests & rebase with latest test project.

aljones15 committed:
Update tests for ttl.

aljones15 committed:
Revise createSession test to have no ttl.

aljones15 committed:
Use branches.

aljones15 committed:
Comment test and use expectedTTL for delay.

aljones15 committed:
Emit storage event to sync tabs.

aljones15 committed:
Re-add End + extend wait times.

aljones15 committed:
Use expectedTTL to ensure sessions expire between tests.

aljones15 committed:
Add another logout step.

aljones15 committed:
Use a json object for localStorage.

mattcollier commented:
Requires: https://github.com/digitalbazaar/bedrock-session-http/pull/2

mattcollier commented:
@aljones15 is this PR still relevant? Is this code/branch being used somewhere (without disclosing details)?

aljones15 commented:
@mattcollier this PR is not being used anywhere else, but expiration is still a relevant problem. This PR could be updated in the future.
